icicle/include/hash/hash.cuh

```diff
@@ -0,0 +1,257 @@
+    {
+    }
+
+    virtual cudaError_t pad_many(
+      Image* states,
+      unsigned int number_of_states,
+      unsigned int input_block_len,
+      unsigned int rate,
+      const device_context::DeviceContext& ctx) const
+    {
+      return cudaError_t::cudaSuccess;
+    };
+
+    virtual cudaError_t squeeze_states(
+      const Image* states,
+      unsigned int number_of_states,
+      unsigned int rate,
+      unsigned int offset,
+      bool align,
+      Image* output,
+      const device_context::DeviceContext& ctx) const = 0;
+
+    virtual cudaError_t run_permutation_kernel(
+      const Image* states, Image* output, unsigned int number_of_states, bool aligned, const device_context::DeviceContext& ctx) const = 0;
+  };
+
+  struct SpongeConfig {
+    device_context::DeviceContext ctx;
+    bool are_inputs_on_device;
+    bool are_outputs_on_device;
+    unsigned int input_rate;
+    unsigned int output_rate;
+    unsigned int offset;
+    bool recursive_squeeze;
+    bool aligned;
+    bool is_async;
+  };
+
+  /**
+   * A function that returns the default value of [SpongeConfig](@ref SpongeConfig) for the [SpongeHasher](@ref
+   * SpongeHasher) class.
+   * @return Default value of [SpongeConfig](@ref SpongeConfig).
+   */
+  static SpongeConfig
+  default_sponge_config(const device_context::DeviceContext& ctx = device_context::get_default_device_context())
+  {
+    SpongeConfig config = {
+      ctx,   // ctx
+      false, // are_inputs_on_device
+      false, // are_outputs_on_device
+      0,     // input_rate
+      0,     // output_rate
+      0,     // offset
+      false, // recursive_squeeze
+      false, // recursive_squeeze
+      false, // is_async
+    };
+    return config;
+  }
+
+  template <typename H, typename PreImage, typename Image>
+  class SpongeHasher
+  {
+  public:
+    unsigned int get_width() const {
+      return static_cast<const H*>(this)->width;
+    }
+
+    unsigned int get_preimage_max_length() const {
+      return static_cast<const H*>(this)->preimage_max_length;
+    }
+
+    cudaError_t allocate_states(Image** states, unsigned int number_of_states, const device_context::DeviceContext& ctx) const
+    {
+      CHK_INIT_IF_RETURN();
+      return cudaMallocAsync(states, number_of_states * get_width() * sizeof(Image), ctx.stream);
+    }
+
+    cudaError_t free_states(Image* states, const device_context::DeviceContext& ctx) const
+    {
+      CHK_INIT_IF_RETURN();
+      return cudaFreeAsync(states, ctx.stream);
+    }
+
+    cudaError_t absorb_many(
+      const PreImage* inputs,
+      Image* states,
+      unsigned int number_of_states,
+      unsigned int input_block_len,
+      const SpongeConfig& cfg) const
+    {
+      CHK_INIT_IF_RETURN();
+      unsigned int width = get_width();
+      unsigned int preimage_max_length = get_preimage_max_length();
+
+      if (cfg.input_rate * sizeof(PreImage) > preimage_max_length * sizeof(Image))
```

Comment:
(1) max_length is mix #elements? assuming posseidon.
(2) for keccak what is it? and what the Image type would be?

Reply:
1) max length refers to domain separation. For poseidon2 and keccak it is set to be equal to width
2) Image type is u64